receiver/scraper.go

```diff
@@ -0,0 +1,120 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expvarreceiver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/expvarreceiver/internal/metadata"
+)
+
+type expVar struct {
+	// Use the existing runtime struct for decoding the JSON.
+	MemStats *runtime.MemStats `json:"memstats"`
+}
+
+type expVarScraper struct {
+	cfg        *Config
+	set        *component.ReceiverCreateSettings
+	httpClient *http.Client
+	mb         *metadata.MetricsBuilder
+}
+
+func newExpVarScraper(cfg *Config, set component.ReceiverCreateSettings) *expVarScraper {
+	return &expVarScraper{
+		cfg: cfg,
+		set: &set,
+		mb:  metadata.NewMetricsBuilder(cfg.MetricsConfig),
+	}
+}
+
+func (e *expVarScraper) start(_ context.Context, host component.Host) error {
+	httpClient, err := e.cfg.HTTPClientSettings.ToClient(host.GetExtensions(), e.set.TelemetrySettings)
+	if err != nil {
+		return err
+	}
+	e.httpClient = httpClient
+	return nil
+}
+
+func (e *expVarScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	emptyMetrics := pmetric.NewMetrics()
+	resp, err := e.httpClient.Get(e.cfg.Endpoint)
+	if err != nil {
+		return emptyMetrics, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return emptyMetrics, fmt.Errorf("expected 200 but received %d status code", resp.StatusCode)
+	}
+
+	result, err := decodeResponseBody(resp.Body)
+	if err != nil {
+		return emptyMetrics, err
+	}
+	memStats := result.MemStats
+	if memStats == nil {
+		return emptyMetrics, fmt.Errorf("unmarshalled memstats data is nil")
+	}
+
+	now := pcommon.NewTimestampFromTime(time.Now())
```

Comment:
A virtual clock would be handy here since you could control when `now` is for unit testing outputs.

"github.com/tilinna/clock" is a good package for this since you can use the method:

```golang
clock.FromContext(ctx).Now()
```

Reply:
That is handy to know for future, but I'm unsure it is the right place to use it here. This is because the metrics validation in the tests is done through the internal `scrapertest` package which doesn't check the timestamps. If we ever do want to validate timestamps in tests I can switch over.

For now, I've followed how other similar scraper components are doing timestamps.